vowpalwabbit/cb_explore_adf_squarecb.cc

```diff
@@ -0,0 +1,348 @@
+  _max_costs.resize(num_actions);
+
+  _ex_as.clear();
+  _ex_costs.clear();
+
+  // backup cb example data
+  for (const auto& ex : examples)
+  {
+    _ex_as.push_back(ex->pred.a_s);
+    _ex_costs.push_back(ex->l.cb.costs);
+  }
+
+  // set regressor predictions
+  for (const auto& as : _ex_as[0])
+  {
+    examples[as.action]->pred.scalar = as.score;
+  }
+
+  const float cmin = _min_cb_cost;
+  const float cmax = _max_cb_cost;
+
+  for (size_t a = 0; a < num_actions; ++a)
+  {
+    example* ec = examples[a];
+    ec->l.simple.label = cmin - 1;
+    float sens = base.sensitivity(*ec);
+    float w = 0;  // importance weight
+
+    if (ec->pred.scalar < cmin || std::isnan(sens) || std::isinf(sens))
+      _min_costs[a] = cmin;
+    else
+    {
+      w = binary_search(ec->pred.scalar - cmin + 1, delta, sens);
+      _min_costs[a] = (std::max)(ec->pred.scalar - sens * w, cmin);
+      if (_min_costs[a] > cmax)
+        _min_costs[a] = cmax;
+    }
+
+    if (!min_only)
+    {
+      ec->l.simple.label = cmax + 1;
+      sens = base.sensitivity(*ec);
+      if (ec->pred.scalar > cmax || std::isnan(sens) || std::isinf(sens))
+      {
+        _max_costs[a] = cmax;
+      }
+      else
+      {
+        w = binary_search(cmax + 1 - ec->pred.scalar, delta, sens);
+        _max_costs[a] = (std::min)(ec->pred.scalar + sens * w, cmax);
+        if (_max_costs[a] < cmin)
+          _max_costs[a] = cmin;
+      }
+    }
+  }
+
+  // reset cb example data
+  for (size_t i = 0; i < examples.size(); ++i)
+  {
+    examples[i]->pred.a_s = _ex_as[i];
+    examples[i]->l.cb.costs = _ex_costs[i];
+  }
+}
+
+template <bool is_learn>
+void cb_explore_adf_squarecb::predict_or_learn_impl(VW::LEARNER::multi_learner& base, multi_ex& examples)
+{
+  if (is_learn)
+  {
+    for (size_t i = 0; i < examples.size() - 1; ++i)
+    {
+      CB::label& ld = examples[i]->l.cb;
+      if (ld.costs.size() == 1)
+        ld.costs[0].probability = 1.f;  // no importance weighting
+    }
+
+    VW::LEARNER::multiline_learn_or_predict<true>(base, examples, examples[0]->ft_offset);
+    ++_counter;
+  }
+  else
+    VW::LEARNER::multiline_learn_or_predict<false>(base, examples, examples[0]->ft_offset);
+
+  v_array<ACTION_SCORE::action_score>& preds = examples[0]->pred.a_s;
+  uint32_t num_actions = (uint32_t)preds.size();
+
+  // The actual parameter $\gamma$ used in the SquareCB.
+  const float gamma = _gamma_scale*pow(_counter, _gamma_exponent); 
+
+  // RegCB action set parameters
+  const float max_range = _max_cb_cost - _min_cb_cost;
+  // threshold on empirical loss difference
+  const float delta = _c0 * log((float)(num_actions * _counter)) * pow(max_range, 2);
+
+  // SquareCB Exploration
+  if (!is_learn)
+    {
+      if (!_elim) // Vanilla variant (perform SquareCB exploration over all actions)
+	{
+	  size_t a_min = 0;
+	  float min_cost = preds[0].score;
+	  // Compute highest-scoring action
+	  for(size_t a = 0; a < num_actions; ++a)
+	    {
```

Comment:
Seems like the formatting is a bit off here, running `clang-format` on this file should fix it.